server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+                                    logger.info("decommission request for attribute [{}] failed", decommissionAttribute.toString());
+                                    DecommissionAttributeMetadata decommissionAttributeMetadata = new DecommissionAttributeMetadata(
+                                        decommissionAttribute,
+                                        DecommissionStatus.DECOMMISSION_FAILED
+                                    );
+                                    mdBuilder.putCustom(DecommissionAttributeMetadata.TYPE, decommissionAttributeMetadata);
+                                    return ClusterState.builder(currentState).metadata(mdBuilder).build();
+                                }
+
+                                @Override
+                                public void onFailure(String source, Exception e) {
+                                    logger.error(() -> new ParameterizedMessage(
+                                        "failed to mark status as DECOMMISSION_FAILED for decommission attribute [{}]",
+                                        decommissionAttribute.toString()), e);
+//                                    listener.onFailure(e);
+                                }
+                            }
+                        );
+                    }
+                    listener.onFailure(e);
+                }
+
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                    if (!newState.equals(oldState)) {
+                        // TODO - drain the nodes before decommissioning
+                        failDecommissionedNodes(newState);
+                        listener.onResponse(new ClusterStateUpdateResponse(true));
+                    } else {
+                        listener.onResponse(new ClusterStateUpdateResponse(false));
+                    }
+                }
+            }
+        );
+    }
+
+    private static void validateAwarenessAttribute(final DecommissionAttribute decommissionAttribute, List<String> awarenessAttributes) {
+        if (!awarenessAttributes.contains(decommissionAttribute.attributeName())) {
+            throw new DecommissionFailedException(decommissionAttribute, "invalid awareness attribute requested for decommissioning");
+        }
+        // TODO - should attribute value be part of force zone values? If yes, read setting and throw exception if not found
+    }
+
+    private static void ensureNoAwarenessAttributeDecommissioned(
+        DecommissionAttributeMetadata decommissionAttributeMetadata,
+        DecommissionAttribute decommissionAttribute
+    ) {
+        // If the previous decommission request failed, we will allow the request to pass this check
+        if (decommissionAttributeMetadata != null && !decommissionAttributeMetadata.status().equals(DecommissionStatus.DECOMMISSION_FAILED)) {
```

Comment:
should we check the previous decommission request was for same attribute ?

Reply:
Are you saying if the same requested attribute had a failure in decommissioning previously, we should fail this new request early on?

Reply:
Added checks